File: Typescript/src/DelegationSortedIntList.ts
/**
 * DelegationSortedIntList -- a variant of a SortedIntList that keeps
 * count of the number of attempted element insertions (not to be confused
 * with the current size, which goes down when an element is removed)
 * and exports an accessor (totalAdded) for this count.
 *
 * @author Nora Shoemaker
 *
 */

import { IntegerList } from "./IntegerList";
import { SortedIntList } from "./hidden/SortedIntListLibrary";

class DelegationSortedIntList implements IntegerList {
  private sortedIntList: SortedIntList;
  private totalAdded: number = 0;

  /**
   * Constructs a new DelegationSortedIntList.
   */
  constructor() {
    this.sortedIntList = new SortedIntList();
  }

  /**
   * Adds an integer to the list.
   *
   * @param num The integer to be added.
   * @returns True if the integer is added successfully, false otherwise.
   */
  add(num: number): boolean {
    this.totalAdded++;
    return this.sortedIntList.add(num);
  }

  /**
   * Adds all integers from the provided IntegerList to the list.
   *
   * @param list The IntegerList containing elements to be added.
   * @returns True if any element is added successfully, false otherwise.
   */
  addAll(list: IntegerList): boolean {
    let added = false;
    for (let i = 0; i < list.size(); i++) {
      added = this.add(list.get(i)) || added;
    }
    return added;
  }

  /**
   * Retrieves the integer at the specified position in the list.
   *
   * @param index The index of the element to retrieve.
   * @returns The element at the specified position in the list.
   */
  get(index: number): number {
    return this.sortedIntList.get(index);
  }

  /**
   * Removes the specified integer from the list.
   *
   * @param num The integer to be removed.
   * @returns True if the integer is removed successfully, false otherwise.
   */
  remove(num: number): boolean {
    return this.sortedIntList.remove(num);
  }

  /**
   * Removes all integers from the provided IntegerList from the list.
   *
   * @param list The IntegerList containing elements to be removed.
   * @returns True if any element is removed successfully, false otherwise.
   */
  removeAll(list: IntegerList): boolean {
    return this.sortedIntList.removeAll(list);
  }

  /**
   * Retrieves the size of the list.
   *
   * @returns The number of elements in the list.
   */
  size(): number {
    return this.sortedIntList.size();
  }

  /**
   * Retrieves the total number of elements added to the list.
   *
   * @returns The total number of elements added to the list.
   */
  getTotalAdded(): number {
    return this.totalAdded;
  }
}

export { DelegationSortedIntList }

File: Java/src/main/java/edu/cmu/cs/cs214/rec04/DelegationSortedIntList.java
package edu.cmu.cs.cs214.rec04;

/**
 * DelegationSortedIntList -- a variant of a SortedIntList that keeps
 * count of the number of attempted element insertions (not to be confused
 * with the current size, which goes down when an element is removed)
 * and exports an accessor (totalAdded) for this count.
 *
 * @author Nora Shoemaker
 *
 */

public class DelegationSortedIntList implements IntegerList {
    private SortedIntList sortedIntList;
    private int totalAdded;

    /**
     * Constructs a new DelegationSortedIntList.
     * Initializes a SortedIntList instance for delegation and sets totalAdded to 0.
     */
    public DelegationSortedIntList() {
        sortedIntList = new SortedIntList();
        totalAdded = 0;
    }

    /**
     * Adds the specified integer to the list while maintaining ascending order.
     * Delegates the operation to the SortedIntList instance and tracks added elements.
     *
     * @param num an integer to be added to the list
     * @return true if the list is changed as a result of the call
     */
    @Override
    public boolean add(int num) {
        boolean added = sortedIntList.add(num);
        if (added) {
            totalAdded++;
        }
        return added;
    }

    /**
     * Adds all integers from the specified IntegerList to this list while maintaining ascending order.
     * Delegates the operation to the SortedIntList instance and tracks added elements.
     *
     * @param list IntegerList containing elements to be added to the list
     * @return true if the list changed as a result of the call
     */
    @Override
    public boolean addAll(IntegerList list) {
        boolean addedAll = sortedIntList.addAll(list);
        if (addedAll) {
            totalAdded += list.size();
        }
        return addedAll;
    }

    /**
     * Returns the integer at the specified position in this list.
     *
     * @param index index of the element to return
     * @return the element at the specified position in this list
     */
    @Override
    public int get(int index) {
        return sortedIntList.get(index);
    }

    /**
     * Removes the first occurrence of the specified element from the list.
     *
     * @param num an integer to be removed from the list, if present
     * @return true if an element was removed as a result of this call
     */
    @Override
    public boolean remove(int num) {
        return sortedIntList.remove(num);
    }

    /**
     * Removes from the list all of its elements that are contained in the specified IntegerList.
     *
     * @param list IntegerList containing elements to be removed from the list
     * @return true if the list changed as a result of the call
     */
    @Override
    public boolean removeAll(IntegerList list) {
        return sortedIntList.removeAll(list);
    }

    /**
     * Returns the number of elements in this list.
     *
     * @return number of elements in the list
     */
    @Override
    public int size() {
        return sortedIntList.size();
    }

    /**
     * Returns the total count of attempted element insertions since creation.
     *
     * @return the total count of attempted element insertions
     */
    public int getTotalAdded() {
        return totalAdded;
    }
}